Sources/ReactiveArray.swift

```diff
@@ -0,0 +1,258 @@
+        precondition(!isEmpty, "can't remove first element from an empty array")
+        return remove(at: startIndex)
+    }
+
+    public func removeFirst(_ n: Int) {
+        precondition(n >= 0, "number of elements to remove should be non-negative")
+        precondition(n <= count, "can't remove more items from an array than it has")
+        removeSubrange(startIndex..<startIndex.advanced(by: n))
+    }
+
+    public func removeLast() -> Element {
+        precondition(!isEmpty, "can't remove last element from an empty array")
+        return remove(at: index(before: endIndex))
+    }
+
+    public func removeLast(_ n: Int) {
+        precondition(n >= 0, "number of elements to remove should be non-negative")
+        precondition(n <= count, "can't remove more items from an array than it has")
+        removeSubrange(endIndex.advanced(by: -n)..<endIndex)
+    }
+
+    public func removeSubrange(_ bounds: ClosedRange<Int>) {
+        replaceSubrange(bounds, with: EmptyCollection())
+    }
+
+    public func removeSubrange(_ bounds: CountableRange<Int>) {
+        replaceSubrange(bounds, with: EmptyCollection())
+    }
+
+    public func removeSubrange(_ bounds: CountableClosedRange<Int>) {
+        replaceSubrange(bounds, with: EmptyCollection())
+    }
+
+    public func removeSubrange(_ bounds: Range<Int>) {
+        replaceSubrange(bounds, with: EmptyCollection())
+    }
+
+    public func replaceSubrange<C>(_ subrange: ClosedRange<Int>, with newElements: C) where C : Collection, C.Iterator.Element == Element {
+        replaceSubrange(Range(subrange), with: newElements)
+    }
+
+    public func replaceSubrange<C>(_ subrange: CountableRange<Int>, with newElements: C) where C : Collection, C.Iterator.Element == Element {
+        replaceSubrange(Range(subrange), with: newElements)
+    }
+
+    public func replaceSubrange<C>(_ subrange: CountableClosedRange<Int>, with newElements: C) where C : Collection, C.Iterator.Element == Element {
+        replaceSubrange(Range(subrange), with: newElements)
+    }
+
+    public func replaceSubrange<C>(_ subrange: Range<Int>, with newElements: C) where C: Collection, C.Iterator.Element == Element {
+
+        let changeset = Changeset(
+            deletions:  changes(removing: elements[subrange], at: subrange),
+            insertions: changes(inserting: newElements, at: subrange)
+        )
+
+        elements.replaceSubrange(subrange, with: newElements)
+
+        innerObserver.send(value: changeset)
+    }
+
+    public func reserveCapacity(_ n: Int) {
+        elements.reserveCapacity(n)
+    }
+
+}
+
+// MARK: - Helpers
+
+extension ReactiveArray {
+
+    fileprivate func changes<C>(removing elements: C, at subrange: Range<Int>) -> [Remove<Element>] where C: Collection, C.Iterator.Element == Element {
+        return elements
+            .enumerated()
+            .map { ($0.advanced(by: subrange.lowerBound), $1) }
+            .map(flip(Remove.init))
```

Comment:
It might be nice to guarantee that removal indexes are emitted in descending order.